Simplify CacheAllocatorConfig implementation for memory tiers

igchor opened the pull request:
Revert some of the changes made earlier and reimplement them without
chaning getMemoryTierConfigs().

After this change, it is no longer possible to call setCacheSize()
and configureMemoryTiers() in arbitrary order - setCacheSize() must
be called first.

I'm not 100% sure if this way is better because if we introduce some new global setting which we would like to apply for all tiers (like currently usePosixShm) we will have to do that both in configureMemoryTiers and in function setting this property (see usePosixShm for example). Currently, we can do that just once, in getMemoryTierConfigs()

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/cachelib/13)
<!-- Reviewable:end -->

igchor committed:
Adjust and enable tests for ShmFileSegment

igchor committed:
Introduce FileShmSegment for file-backed shared memory

It's implementation is mostly based on PosixShmSegment.

Also, extend ShmManager and ShmSegmentOpts to support this new
segment type.

guptask committed:
Add support for shm opts serialization

After introducing file segment type, nameToKey_ does not provide
enough information to recover/remove segments on restart.

This commit fixes that by replacing nameToKey_ with nameToOpts_.

Previously, the Key from nameToKey_ map was only used in a single
DCHECK().

victoria-mcgrath committed:
Initial version of config API extension to support multiple memory tiers

* New class MemoryTierCacheConfig allows to configure a memory tier.
  Setting tier size and location of a file for file-backed memory are
  supported in this initial implementation;
* New member, vector of memory tiers, is added to class CacheAllocatorConfig.
* New test suite, chelib/allocator/tests/MemoryTiersTest.cpp,
  demonstrates the usage of and tests extended config API.

igchor committed:
Integrate Memory Tier config API with CacheAllocator.

igchor committed:
Run centos and debian workflows on push and PR

igchor committed:
Merge pull request #4 from igchor/workflows_pr

Run centos and debian workflows on push and PR

guptask committed:
Merge pull request #3 from igchor/develop2

Publish changes from innersouce

igchor committed:
Run tests on CI

igchor committed:
Add MemoryTierCacheConfig::fromShm()

to allow using new configureMemoryTiers() API with legacy behavior.

Move validation code for memory tiers to validate() method and convert
ratios to sizes lazily (on get)..

guptask committed:
Moved common segment code for posix and file shm segments into ShmCommon

igchor committed:
Fix test_shm_manager.cpp test

It wrongly assumed that the only possible segment type is
PosixSysV segment.

igchor committed:
Merge pull request #2 from igchor/config_api_integration

Config api integration

igchor committed:
Merge pull request #6 from igchor/enable_tests

Run tests on CI

igchor committed:
Run long tests (navy/bench) every day on CI

victoria-mcgrath committed:
Enabled memory tier config API for cachebench.

igchor committed:
Merge pull request #7 from guptask/segment_common_migration

FBSW-208 Moved common segment code for posix and file shm segments into ShmCommon

igchor committed:
Merge pull request #10 from igchor/cron_tests

Run long tests (navy/bench) every day on CI

victoria-mcgrath commented:
Honestly, I like this version better because it seems cleaner and safer. I do agree that dealing with cache-global settings is the problem here. What could we do instead of propagating every global setting to every tier? Would it make more sense to create a kind of property container for cache-wide settings which tiers could reference?





<!-- Sent from Reviewable.io -->

igchor committed:
Simplify CacheAllocatorConfig implementation for memory tiers

Revert some of the changes made earlier and reimplement them without
chaning getMemoryTierConfigs().

After this change, it is no longer possible to call setCacheSize()
and configureMemoryTiers() in arbitrary order - setCacheSize() must
be called first.

igchor reviewed (commented):
Right, this might be one possible approach - but I'm not sure if we should do this now. If you like this solution then I'm OK with merging it as soon as tests will pass.

> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/13#-:-MpCW8cZ3iB-JdUi19YR:b-gc6c4d)* status: 0 of 2 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

victoria-mcgrath reviewed (approved):
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/13)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @igchor)



<!-- Sent from Reviewable.io -->
